x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/Expression.java

```diff
@@ -114,7 +114,15 @@ public final Expression canonical() {
     }
 
     protected Expression canonicalize() {
-        return this;
+        if (children().isEmpty()) {
+            return this;
+        }
+        List<Expression> canonicalChildren = Expressions.canonicalize(children());
+        // check if replacement is really needed
+        if (children().equals(canonicalChildren)) {
```

Comment:
If it's possible, for the future, maybe we should move this handling inside the `replaceChildren()`